tests/test_lion8b.py

```diff
@@ -0,0 +1,422 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import time
+import warnings
+
+import numpy as np
+import pytest
+import torch
+
+from llmfoundry.optim import DecoupledLionW_8bit as Lion8bit
+
+warnings.filterwarnings('ignore')
+
+_MANY_PARAM_SHAPES = [(1, 1), (1, 2), (17, 23), (64, 32)]
+_FLOAT_DTYPES = [torch.bfloat16, torch.float16, torch.float32]
+
+np.set_printoptions(linewidth=160, formatter={'float': lambda f: f'{f:5.3f}'})
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('N,D', _MANY_PARAM_SHAPES)
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+@pytest.mark.parametrize('fused,use_errors', [(False, False), (True, False),
+                                              (True, True)])
+def test_modifies_weights_and_momentums(N: int, D: int, dtype: torch.dtype,
+                                        fused: bool, use_errors: bool) -> None:
+    device = 'cuda'
+    torch.manual_seed(123)
+    X = torch.randn((N, D), device=device, requires_grad=False, dtype=dtype)
+    W = torch.randn((D, D), device=device, requires_grad=True, dtype=dtype)
+    W_orig = W.detach().clone()
+
+    opt = Lion8bit([W],
+                   lr=1.0,
+                   _fused=fused,
+                   betas=(.75, .75),
+                   weight_decay=.2,
+                   error_correction=use_errors)
+
+    Y = X @ W
+    loss = Y.sum()
+    loss.backward()
+    torch.testing.assert_close(W_orig, W)  # no weight modification yet
+    opt.step()
+    opt.zero_grad()
+
+    with pytest.raises(AssertionError):  # opt step modified the weights
+        torch.testing.assert_close(W_orig, W)
+
+    # every momentum should be nonzero with infinite precision, but
+    # might be zero after quantization
+    param_state = opt.state[W]  # type:ignore using tensor as key
+    momentum = param_state['exp_avg'].materialize()
```

Comment:
Is there a `torch.Tensor.materialize()` member method?
Can't find it in the docs... can this be replaced with something people can reference?

Reply:
Negative--this is specific to our MaybeQuantizedTensor. We make sure that our exp_avg state is always an instance of this class.

I'll add some clarifying comments.